pkg/reader/usdc_reader.go

```diff
@@ -0,0 +1,307 @@
+		}
+		e, err1 := event.unpackID()
+		if err1 != nil {
+			return nil, err1
+		}
+		messageSentEvents[e] = event.Arg0
+	}
+
+	// 3. This should be done by ChainReader - picking only events matching eventIDs.
+	// Right now ChainReader doesn't support filtering by specific data words
+	out := make(map[exectypes.MessageTokenID]MessageHash)
+	for tokenID, messageID := range eventIDs {
+		messageHash, ok1 := messageSentEvents[messageID]
+		if !ok1 {
+			// Token not available in the source chain
+			continue
+		}
+		out[tokenID] = messageHash
+	}
+
+	return out, nil
+}
+
+func (u usdcMessageReader) recreateMessageTransmitterEvents(
+	destChainSelector cciptypes.ChainSelector,
+	tokens map[exectypes.MessageTokenID]cciptypes.RampTokenAmount,
+) (map[exectypes.MessageTokenID]eventID, error) {
+	messageTransmitterEvents := make(map[exectypes.MessageTokenID]eventID)
+
+	for id, token := range tokens {
+		sourceTokenPayload, err := NewSourceTokenDataPayloadFromBytes(token.ExtraData)
+		if err != nil {
+			return nil, err
+		}
+
+		// USDC message payload:
+		// uint32 _msgVersion,
+		// uint32 _msgSourceDomain,
+		// uint32 _msgDestinationDomain,
+		// uint64 _msgNonce,
+		// bytes32 _msgSender,
+		// Since it's packed, all of these values contribute to the first slot
+
+		msgVersionBytes := [4]byte{}
+		binary.BigEndian.PutUint32(msgVersionBytes[:], CCTPMessageVersion)
+
+		sourceDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(sourceDomainBytes[:], sourceTokenPayload.SourceDomain)
+
+		destDomain, ok := u.cctpDestDomain[uint64(destChainSelector)]
+		if !ok {
+			return nil, fmt.Errorf("destination domain not found for chain %s", destChainSelector)
+		}
+
+		destDomainBytes := [4]byte{}
+		binary.BigEndian.PutUint32(destDomainBytes[:], destDomain)
+
+		nonceBytes := [8]byte{}
+		binary.BigEndian.PutUint64(nonceBytes[:], sourceTokenPayload.Nonce)
+
+		senderBytes := [12]byte{}
+
+		var buf bytes.Buffer
+		buf.Write(msgVersionBytes[:])
+		buf.Write(sourceDomainBytes[:])
+		buf.Write(destDomainBytes[:])
+		buf.Write(nonceBytes[:])
+		buf.Write(senderBytes[:])
+		messageTransmitterEvents[id] = [32]byte(buf.Bytes()[:32])
+	}
+	return messageTransmitterEvents, nil
+}
+
+func bindMessageTransmitters(
+	ctx context.Context,
+	readers map[cciptypes.ChainSelector]contractreader.ContractReaderFacade,
+	chainSel cciptypes.ChainSelector,
+	address string,
+) (types.BoundContract, error) {
+	var empty types.BoundContract
+	r, ok := readers[chainSel]
+	if !ok {
+		return empty, fmt.Errorf("no contract reader found for chain %d", chainSel)
+	}
+
+	contract := types.BoundContract{
+		Address: address,
+		Name:    consts.ContractNameCCTPMessageTransmitter,
+	}
+	if err := r.Bind(ctx, []types.BoundContract{contract}); err != nil {
+		return empty,
+			fmt.Errorf(
+				"unable to bind %s for chain %d: %w",
+				consts.ContractNameCCTPMessageTransmitter, chainSel, err,
+			)
+	}
+
+	return contract, nil
+}
```

Comment:
This is nearly identical to the generalized version `bindReaderContract` in `ccipChainReader`. Maybe worth unifying?

One difference is that the address in that version is the []byte version rather than the encoded version.